File: ipification-sdk-android-java/app/src/main/java/com/ipification/demoapp/manager/IPHelper.java
package com.ipification.demoapp.manager;


import android.util.Log;

import com.ipification.demoapp.BuildConfig;
import com.ipification.demoapp.Urls;
import com.ipification.demoapp.callback.TokenCallback;
import com.ipification.demoapp.util.Util;
import com.ipification.mobile.sdk.android.IPConfiguration;

import org.jetbrains.annotations.NotNull;

import java.io.IOException;

import okhttp3.Call;
import okhttp3.Callback;
import okhttp3.FormBody;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;

public class IPHelper {

    private static final String TAG = "IPHelper";


    public static void doPostToken(String code, final TokenCallback callback) {
        try {
            String url = Urls.getExchangeTokenUrl();

            RequestBody body = new FormBody.Builder()
                    .add("client_id", IPConfiguration.getInstance().getCLIENT_ID())
                    .add("redirect_uri", IPConfiguration.getInstance().getREDIRECT_URI().toString())
                    .add("grant_type", "authorization_code")
                    .add("client_secret", BuildConfig.CLIENT_SECRET)
                    .add("code", code)
                    .build();

            OkHttpClient client = new OkHttpClient.Builder().build();
            Request request = new Request.Builder().url(url).post(body)
                    .build();
            Call call = client.newCall(request);
            call.enqueue(new Callback() {
                @Override
                public void onResponse(@NotNull Call call, @NotNull Response response) throws IOException {
                    if (response.isSuccessful() && response.body() != null) {
                        String accessToken = Util.parseAccessTokenFromJSON(response.body().string());
                        if (!accessToken.equals("")) {
                            // call userInfo
                            doPostUserInfo(accessToken, callback);
                        } else {
//                            callback.result("", response.body().string());
                        }
                    } else {
                        try {
                            if (response.body() != null) {
//                                callback.result("", response.body().string());
                            } else {
//                                callback.result("", "error body is null " + response.code());
                            }
                        } catch (Exception e) {
                            e.printStackTrace();
                        }
                    }
                }

                @Override
                public void onFailure(@NotNull Call call, @NotNull IOException e) {
//                    callback.result("", e.getMessage());
                    Log.e(TAG, "token exchange error: " + e.getMessage());
                }
            });
        } catch (Exception e) {
            Log.e(TAG, "token exchange error: " + e.getMessage());
            e.printStackTrace();
//            callback.result("", e.getMessage());
        }
    }
    public static void doPostUserInfo(String accessToken, final TokenCallback callback) {
        try {
            String url = Urls.getUserInfoUrl();

            RequestBody body = new FormBody.Builder()
                    .add("access_token", accessToken)
                    .build();

            OkHttpClient client = new OkHttpClient.Builder().build();
            Request request = new Request.Builder().url(url).post(body)
                    .build();
            Call call = client.newCall(request);
            call.enqueue(new Callback() {
                @Override
                public void onResponse(@NotNull Call call, @NotNull Response response) throws IOException {
                    if (response.isSuccessful() && response.body() != null) {
//                        callback.result(response.body().string(), "");
                    } else {
                        try {
                            if (response.body() != null) {
//                                callback.result("", response.body().string());
                            } else {
//                                callback.result("", "error body is null " + response.code());
                            }
                        } catch (Exception e) {
                            e.printStackTrace();
                        }
                    }
                }

                @Override
                public void onFailure(@NotNull Call call, @NotNull IOException e) {
//                    callback.result("", e.getMessage());
                    Log.e(TAG, "token exchange error: " + e.getMessage());
                }
            });
        } catch (Exception e) {
            Log.e(TAG, "token exchange error: " + e.getMessage());
            e.printStackTrace();
//            callback.result("", e.getMessage());
        }
    }
}

File: ipification-sdk-android-kotlin/app/src/main/java/com/ipification/demoapp/manager/IPHelper.kt
package com.ipification.demoapp.manager;

import android.app.Activity
import android.util.Log
import com.ipification.demoapp.Urls
import com.ipification.demoapp.callback.TokenCallback
import com.ipification.demoapp.util.Util
import com.ipification.mobile.sdk.android.IPConfiguration
import okhttp3.Call
import okhttp3.Callback
import okhttp3.FormBody
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody
import okhttp3.Response
import java.io.IOException

class IPHelper {
    companion object Factory {
        private const val TAG = "IPHelper"

        // Just for testing. The client app needs to call their backend service.
        fun callTokenExchangeAPI(activity: Activity, code: String) {
            doPostToken(code, object : TokenCallback {
                override fun onSuccess(response: String) {
                    handleTokenExchangeSuccess(activity, response)
                }

                override fun onError(errorMessage: String) {
                    Util.openErrorActivity(activity, errorMessage ?: "unknown error")
                }
            })
        }

        private fun handleTokenExchangeSuccess(activity: Activity, response: String) {
            val phoneNumberVerified = Util.parseTokenJSON(response, "phone_number_verified")
            val phoneNumber = Util.parseTokenJSON(response, "phone_number")

            if (phoneNumberVerified == "true" || phoneNumber != null) {
                Util.openSuccessActivity(activity, response)
            } else {
                Util.openErrorActivity(activity, response)
            }
        }

        private fun doPostToken(code: String, callback: TokenCallback) {
            val url = Urls.getTokenExchangeUrl()
            val body: RequestBody = FormBody.Builder()
                .add("client_id", IPConfiguration.getInstance().CLIENT_ID)
                .add("grant_type", "authorization_code")
                .add("redirect_uri", IPConfiguration.getInstance().REDIRECT_URI.toString())
                .add("code", code)
                .build()

            val client = OkHttpClient.Builder().build()
            val request: Request = Request.Builder().url(url).post(body).build()

            client.newCall(request).enqueue(object : Callback {
                override fun onFailure(call: Call, e: IOException) {
                    val error = e.localizedMessage
                    Log.e(TAG, "doPostToken error: $error")
                    callback.onError(e.localizedMessage ?: "doPostToken - onFailure")
                }

                override fun onResponse(call: Call, response: Response) {
                    try {
                        val responseBody = response.body?.string()
                        Log.d(TAG, "doPostToken response: $responseBody")

                        if (responseBody != null && response.isSuccessful) {

                            callback.onSuccess(responseBody)
                        } else {
                            callback.onError("doPostToken - error $responseBody")
                        }
                    } catch (e: Exception) {
                        e.printStackTrace()
                        callback.onError("doPostToken - error ${e.message}")
                    }
                }
            })
        }

    }
}

